ocean/helium/snapshots.py

```diff
@@ -131,10 +136,267 @@ def download_file_from_snapshot(src, dst, snapshothash):
+        Returns clone of this package.
+        """
+        return Package(copy.deepcopy(self._data), copy.deepcopy(self._meta))
+
+    def __contains__(self, logical_key):
+        """
+        Checks whether the package contains a specified logical_key.
+
+        Returns:
+            True or False
+        """
+        return logical_key in self._data
+
+    @staticmethod
+    def load(path):
+        """
+        Loads a package from a path.
+
+        Args:
+            path: string representing the location to load the package from
+
+        Returns:
+            a new package object
+
+        Raises:
+            file not found
+            json decode error
+            invalid package exception
+        """
+        data = {}
+        with jsonlines.open(path) as reader:
+            meta = reader.read()
+            for obj in reader:
+                lk = obj['logical_key']
+                if lk in data:
+                    raise PackageException("Duplicate logical key while loading package")
+                del obj['logical_key']
+                data[lk] = obj
+
+        return Package(data, meta)
+
+    @staticmethod
+    def pkg(path):
+        """
+        Takes a package of a provided path.
+
+        Recursively enumerates every file in path, and returns a new
+        package that contains all those files.
+
+        Args:
+            path: path to package
+
+        Returns:
+            A new package of that path
+
+        Raises:
+            when path doesn't exist
+        """
+        # TODO: anything but local paths
+        # TODO: deserialization metadata
+        data = {}
+        meta = {}
+        src_path = pathlib.Path(path)
+        files = src_path.rglob('*')
+        for f in files:
+            if not f.is_file():
+                continue
+            entry = {
+                'hash': hash_file(f),
+                'size': os.path.getsize(f),
+                'user_meta': {},
+                'physical_keys': [{
+                    'type': 'local_file',
+                    'path': os.path.abspath(f)
+                }]
+            }
+            data[str(f)] = entry
+        return Package(data, meta)
+
+    def get(self, logical_key):
+        """
+        Gets object from local_key and returns it as an in-memory object.
+
+        Args:
+            logical_key: logical key of the object to get
+
+        Returns:
+            A deserialized object from the logical_key or
+                a stream of bytes if deserialization info is missing
+
+        Raises:
+            KeyError: when logical_key is not present in the package
+            physical key failure
+            hash verification fail
+        """
+        entry = self._data[logical_key]
+        physical_keys = entry['physical_keys']
+        physical_key = physical_keys[0] # TODO: support multiple physical keys
+        stream = dereference_physical_key(physical_key)
+        # TODO: verify hash
+        if 'target' in entry:
```

Comment:
`stream` is now an open file descriptor... and it'll get leaked if anything happens after this.